cilium: various misc refactoring & improvements

borkmann opened the pull request:
(see commit msg)

borkmann committed:
cilium, bpf: Move lb-only out of datapath mode

It's not a good fit, since this is really about veth & netkit only, and
lb-only mode is misplaced here. Move it into a separate flag which later
can be deprecated.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: Lift watchdog limitation

After the rework in c508c64fd6a3 ("pkg/datapath: Table based loader
reconciliation") the original issue seems to be gone in that the watchdog
triggered due to missing BPF programs on the tc layer.

Alternative option is --endpoint-bpf-prog-watchdog-interval=0s flag.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: Leave BPF regen restrictions only to dry mode

There is no good reason that this needs to be gated behind the LB switch,
thus get rid of these conditionals.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: Lift restoreServicesLocked maglev conditional

Commit 2b9485df05ec ("service: Restore Maglev table when M changes") has
further details on the Maglev logic in combination with M changes. As per
Cilium docs:

  [...] maglev.tableSize specifies the size of the Maglev lookup table
  for each single service. Maglev recommends the table size (M) to be
  significantly larger than the number of maximum expected backends (N).
  In practice that means that M should be larger than 100 * N in order
  to guarantee the property of at most 1% difference in the reassignments
  on backend changes. M must be a prime number. Cilium uses a default
  size of 16381 for M. [...]

For the case where M changes, a disruption should be expected. There is no
particular reason this needs to be gated behind LB knob. Add a note to the
docs on M.

For the unit test, only store the active counter in the fake Maglev table
since it doesn't handle weights internally right now.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, ipam: Use clientset in newIPAMAPIHandler

The IPAM hooks really depend on whether k8s client is enabled or not.
Thus convert to the latter rather than using the LB switch to set up
ipamAPIHandler.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

qmonnet commented on Documentation/network/kubernetes/kubeproxy-free.rst:
```diff
@@ -492,7 +492,10 @@ The following sizes for ``M`` are supported as ``maglev.tableSize`` Helm option:
+difference in reassignments on backend changes will increase. Note that changing the table
+size (``M``) triggers a recalculation of the lookup table and can temporarily lead to inconsistent
+backend selection for new traffic until all nodes have converged and completed their agent
+restart respectively bootstrap.
```
I'm struggling with the end of this sentence, what's `their agent restart respectively bootstrap`? Did you mean restart and bootstrap?

borkmann replied:
yes, will fix

qmonnet commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -81,7 +82,7 @@ cilium-agent [flags]
       --controller-group-metrics strings                          List of controller group names for which to to enable metrics. Accepts 'all' and 'none'. The set of controller group names available is not guaranteed to be stable between Cilium versions.
       --crd-wait-timeout duration                                 Cilium will exit if CRDs are not available within this duration upon startup (default 5m0s)
-      --datapath-mode string                                      Datapath mode name (veth, netkit, netkit-l2, lb-only) (default "veth")
+      --datapath-mode string                                      Datapath mode name (veth, netkit, netkit-l2) (default "veth")
```
This probably warrants an upgrade note for whoever was using `--datapath-mode` to setup `lb-only` mode?

borkmann replied:
ack will add

borkmann replied:
done

qmonnet reviewed (changes requested):
Looks good overall, thanks! Two minor comments inline below.

qmonnet reviewed (approved):
Thank you!

mhofstetter commented on cilium-dbg/cmd/service_update.go:
```diff
@@ -205,10 +204,6 @@ func updateService(cmd *cobra.Command, args []string) {
 			Fatalf("Unable to retrieve cilium configuration: empty response")
 		}
-
-		if resp.Status.DatapathMode != datapathOption.DatapathModeLBOnly {
```
Aren't backend weights still not supported for all datapath modes other than `veth` in combination with  `loadbalancer-only`? Or has something changed in the meantime and this is also supported for other datapath modes and we can really remove it completely?

```
option.Config.DatapathMode != datapathOption.DatapathModeVeth || !option.Config.LoadBalancerOnly
```

and maybe adjusting the error message?

borkmann replied:
This is independent of veth or netkit. The weights support currently only has an exposure through the cilium-dbg cli command line, so you need to configure it manually, if you don't configure it specifically then each backend has equal weight. So removing this conditional here is fine.

mhofstetter commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -167,16 +166,16 @@ func initKubeProxyReplacementOptions(sysctl sysctl.Sysctl, tunnelConfig tunnel.C
-			if option.Config.DatapathMode != datapathOption.DatapathModeLBOnly {
-				return fmt.Errorf("DSR dispatch mode %s only supported for --%s=%s", option.Config.LoadBalancerDSRDispatch, option.DatapathMode, datapathOption.DatapathModeLBOnly)
+			if !option.Config.LoadBalancerOnly {
+				return fmt.Errorf("DSR dispatch mode %s only supported for --%s", option.Config.LoadBalancerDSRDispatch, option.LoadBalancerOnly)
```
`... only supported for --%s=true` ?

mhofstetter commented on pkg/option/config.go:
```diff
@@ -287,6 +287,9 @@ const (
 	LoadBalancerAlgorithmAnnotation = "bpf-lb-algorithm-annotation"
 
+	// LoadBalancerOnly is legacy knob for --datapath-mode=lb-only.
+	LoadBalancerOnly = "bpf-lb-only"
```
is `bpf-lb-only` supported in combination with all datapath modes that are left (veth, netkit, netkitl2) ?

If not, do we need some validation?

borkmann replied:
Yes this is completely independent of it, which is also why this should be moved out of `--datapath-mode`.

mhofstetter commented on pkg/service/service.go:
```diff
@@ -1922,7 +1922,7 @@ func (s *Service) restoreServicesLocked(svcBackendsById map[lb.BackendID]struct{
 		ipv6 := newSVC.frontend.IsIPv6() || (svc.NatPolicy == lb.SVCNatPolicyNat46)
 		recreated := s.lbmap.IsMaglevLookupTableRecreated(ipv6)
-		if option.Config.LoadBalancerOnly && newSVC.useMaglev() && recreated {
+		if newSVC.useMaglev() && recreated {
```
it seems like this change is breaking the tests in  `github.com/cilium/cilium/pkg/service`

```
    service_test.go:1513: 
        	Error Trace:	/home/runner/work/cilium/cilium/pkg/service/service_test.go:1513
        	Error:      	Not equal: 
        	            	expected: 2
        	            	actual  : 3
        	Test:       	TestRestoreServiceWithTerminatingBackends
```

mhofstetter reviewed (commented):
Thanks for the cleanup.

I left some comments that might be worth to address (in addition to fixing the tests)

pippolo84 reviewed (approved):
:100:

derailed reviewed (approved):
@borkmann Nice work Daniel!

mhofstetter commented on daemon/cmd/daemon_main.go:
```diff
@@ -250,9 +250,8 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
-			datapathOption.DatapathModeVeth, datapathOption.DatapathModeNetkit,
-			datapathOption.DatapathModeNetkitL2, datapathOption.DatapathModeLBOnly))
+		fmt.Sprintf("Datapath mode name (%s, %s, %s)",
+			datapathOption.DatapathModeVeth, datapathOption.DatapathModeNetkit, datapathOption.DatapathModeNetkitL2))
```
btw: do we need to cleanup the helm values too? There are at least occurrences in the doc.

https://github.com/cilium/cilium/blob/pr/lb-improvements/install/kubernetes/cilium/values.yaml#L627-L627

And maybe it's also worth to use the new flag `--bpf-lb-only` when `loadbalancer.standalone=true` in the cilium-configmap (https://github.com/cilium/cilium/blob/pr/lb-improvements/install/kubernetes/cilium/templates/cilium-configmap.yaml#L785-L785)?

mhofstetter commented on pkg/metrics/features/metrics.go:
```diff
@@ -164,7 +164,6 @@ var (
 		datapathOption.DatapathModeVeth,
 		datapathOption.DatapathModeNetkit,
 		datapathOption.DatapathModeNetkitL2,
-		datapathOption.DatapathModeLBOnly,
```
Do we need to adjust (re-generate?) the corresponding metrics feature doc too? https://github.com/cilium/cilium/blob/pr/lb-improvements/Documentation/observability/feature-metrics-agent.txt#L311-L311

borkmann committed:
cilium, helm: Update helm documentation and configmap

Update helm documentation and configmap to remove the value from
datapath mode itself.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, docs: Update feature-metrics-agent doc

Remove lb-only mode from the metrics document.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

gandro commented on pkg/ipam/cell/cell.go:
```diff
@@ -70,7 +71,7 @@ type ipamAPIHandlerOut struct {
 
 func newIPAMAPIHandler(params ipamAPIHandlerParams) ipamAPIHandlerOut {
-	if option.Config.LoadBalancerOnly {
+	if !params.Clientset.IsEnabled() {
```
This does not look correct to me. The IPAM API is used even without K8s, e.g. via the docker plugin

borkmann replied:
Thanks for the feedback! @gandro is there any way where we can detect that docker plugin is running to feed this into the above? I want to remove the LoadBalancerOnly conditional ideally. (So only have the empty handlers if neither k8s nor docker plugin is enabled.)

gandro replied:
There might also be other users of the API outside of the docker and CNI plugin. It's an API after all, we don't know who is using it to manually create endpoints.

I think the old approach of disabling it based on a specific flag that the user has to explicitly opt-in is the best appraoch. Though I'm not sure why someone would want to disable it, it shouldn't impact other subsystems after all, since IPAM is still active even if you disable the API.

borkmann committed:
cilium, ipam: Remove clientset conditional in newIPAMAPIHandler

Remove the clientset conditional in newIPAMAPIHandler altogether. Sebastian
mentioned that IPAM API handlers should not be disabled. They are just
forwarding requests from the outside. If k8s is disabled, then cluster-pool
IPAM should still keep working. Normally, the range is derived from k8s,
but in this case it needs to be specified manually via --ipv4-range (see
below). Also, the docker plugin is relying on this and it has k8s client
disabled too. Thus, just remove the conditional altogether to fix it.

Before:

  # ./daemon/cilium-agent --enable-ipv4=true --enable-ipv6=false --datapath-mode=veth \
    --bpf-lb-only=true --bpf-lb-mode=snat --bpf-lb-acceleration=native --devices=enp5s0 \
    --enable-k8s=false
  [...]
  time="2025-01-29T12:55:59.875831476Z" level=info msg="Delete endpoint by containerID request" containerID=8064e0378f25c1e8842cc331fd836516dbf6b0fd4c70798ba03a89302488baed endpointID=793 k8sNamespace=kube-system k8sPodName=coredns-7c65d6cfc9-7glgh subsys=daemon
  time="2025-01-29T12:55:59.875925553Z" level=warning msg="Failed while waiting for initial global identities" endpointID=793 error="initial global identity sync was cancelled: context canceled" subsys=endpoint
  time="2025-01-29T12:55:59.906867027Z" level=info msg="Removed endpoint" ciliumEndpointName=kube-system/coredns-7c65d6cfc9-7glgh containerID=8064e0378f containerInterface=eth0 datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=793 identity=5 ipv4=10.128.205.58 ipv6= k8sPodName=kube-system/coredns-7c65d6cfc9-7glgh subsys=endpoint
  [...]

After:

  # ./daemon/cilium-agent --enable-ipv4=true --enable-ipv6=false --datapath-mode=veth \
    --bpf-lb-only=true --bpf-lb-mode=snat --bpf-lb-acceleration=native --devices=enp5s0 \
    --enable-k8s=false --ipv4-range="10.16.0.0/16"
  [...]
  time="2025-01-29T13:17:13.47010098Z" level=info msg="Delete endpoint by containerID request" containerID=b0adcadf2cd13dd6da8d157e9d0e91fd397ab59e1086f63896c2d156275e0e69 endpointID=2361 k8sNamespace=kube-system k8sPodName=coredns-7c65d6cfc9-7glgh subsys=daemon
  time="2025-01-29T13:17:13.504341657Z" level=info msg="Removed endpoint" ciliumEndpointName=kube-system/coredns-7c65d6cfc9-7glgh containerID=b0adcadf2c containerInterface= datapathPolicyRevision=1 desiredPolicyRevision=1 endpointID=2361 identity=5 ipv4=10.16.124.198 ipv6= k8sPodName=kube-system/coredns-7c65d6cfc9-7glgh subsys=endpoint
  [...]

Keep this here as a separate commit for future reference.

Fixes: e33b36edcd70 ("cilium, ipam: Use clientset in newIPAMAPIHandler")
Suggested-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
